lib/srv/db/postgres/engine.go

```diff
@@ -274,9 +264,44 @@ func (e *Engine) receiveFromClient(client *pgproto3.Backend, server *pgproto3.Fr
 		log.Debugf("Received client message: %#v.", message)
 		switch msg := message.(type) {
 		case *pgproto3.Query:
-			err := e.Audit.OnQuery(e.Context, *sessionCtx, msg.String)
+			// Query message indicates the client is executing a simple query.
+			e.Audit.OnQuery(e.Context, sessionCtx, msg.String)
+		case *pgproto3.Parse:
+			// Parse message is a start of the extended query protocol which
+			// prepares parameterized query for execution. It is never used
+			// by psql, mostly by various GUI clients and programs.
+			//   https://www.postgresql.org/docs/10/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY
+			sessionCtx.Statements.Save(msg.Name, msg.Query)
```

Comment:
this could leak memory if the connection breaks prematurely
the easiest fix is to `defer sessionCtx.Statements.Remove(msg.Name)` here
alternatively, append all statement names to a slice and set a single defer before the loop to clean those out

Reply:
Hm, could you please elaborate on how we can leak memory here, I'm not sure I see it. Statements cache is per-session, so if the connection breaks, the sessionCtx will be eventually garbage collected and the statements cache with it? Unless I'm missing something.

Reply:
Ahh, missed that it's per-session, you're right